app/public/static/js/components/jigsaw-puzzle.js

```diff
@@ -0,0 +1,86 @@
+AFRAME.registerComponent("jigsaw-puzzle", {
+  schema: {
+    jsonData: {
+      parse: JSON.parse,
+      stringify: JSON.stringify,
+    },
+    isSolved: { type: "map" },
+  },
+
+  multiple: true,
+
+  init: function () {
+    const data = this.data.jsonData;
+    this.id = data.id;
+    const puzzleDimension = data.hasOwnProperty("puzzleDimension")
+      ? data.puzzleDimension
+      : 3;
+    // Creates possible x coordinates centred at 0
+    const xCoord = Array.from(Array(puzzleDimension).keys()).map(
+      (x) => (x - (puzzleDimension / 2 - 1 / 2)) * 1.1
+    );
+    // Creates possible y coordinates centred at 0
+    const yCoord = xCoord.slice(0).reverse();
+
+    // Creates possible (x,y) coordinates
+    let coordPairs = yCoord.flatMap((y) => xCoord.map((x) => [x, y]));
+    // Randomize the starting position of the puzzle pieces
+    let randomCoordPairs = coordPairs.slice(0);
+    shuffleArray(randomCoordPairs);
+    let images = [];
+
+    for (i = 0; i < data.images.length; ++i) {
+      let imageData = {};
+      imageData["imageSrc"] = data.images[i];
+      imageData["xTarget"] = coordPairs[i][0];
+      imageData["yTarget"] = coordPairs[i][1];
+      imageData["x"] = randomCoordPairs[i][0];
+      imageData["y"] = randomCoordPairs[i][1];
+
+      images.push(imageData);
+    }
+
+    ordered_puzzle_json = {
+      id: this.id,
+      images: images,
+      useTargets: true,
+      randomizePos: false,
+      scaleBy: data.hasOwnProperty("scaleBy") ? data.scaleBy : 3,
+      is_last_object: "is_last_object" in data ? data.is_last_object : false,
+    };
+
+    this.ordered_puzzle = document.createElement("a-entity");
+    this.ordered_puzzle.setAttribute("id", "ordered-puzzle");
+    this.ordered_puzzle.setAttribute(
+      "ordered-puzzle",
+      "jsonData",
+      JSON.stringify(ordered_puzzle_json)
+    );
+    this.ordered_puzzle.setAttribute("isSolved", this.data.isSolved);
```

Comment:
Instead of adding a static map here, we should add a binding that will automatically call ordered-puzzle's update method:
```
    this.ordered_puzzle.setAttribute(
    "bind__ordered-puzzle",
    "isSolved: solvedObjects"
    );
```
that way, any time `isSolved` is updated, this will automatically call ordered-puzzle's update method. By calling its update method, jigsaw doesn't have to worry about toggling visibility of the "Solved" thingy.

Reply:
Ooh I see, thanks for explaining, updated the PR